spec/lib/diaspora_federation/entities/relayable_spec.rb

```diff
@@ -271,109 +252,283 @@ def verify_signature(pubkey, signature, signed_string)
+        signature_data_with_new_property = "#{author};#{guid};#{parent_guid};#{new_property};#{property}"
+
+        json_hash = Entities::SomeRelayable.new(hash, property_order, "new_property" => new_property).to_json
+        author_signature = json_hash[:entity_data][:author_signature]
+        parent_author_signature = json_hash[:entity_data][:parent_author_signature]
+
+        expect(verify_signature(author_pkey, author_signature, signature_data_with_new_property)).to be_truthy
+        expect(verify_signature(parent_pkey, parent_author_signature, signature_data_with_new_property)).to be_truthy
+      end
+
+      it "doesn't change signatures if they are already set" do
+        json = Entities::SomeRelayable.new(hash_with_fake_signatures).to_json.to_json
+        expect(json).to include_json(entity_data: {author_signature: "aa"})
+        expect(json).to include_json(entity_data: {parent_author_signature: "bb"})
+      end
+
+      it "raises when author_signature not set and key isn't supplied" do
+        expect_callback(:fetch_private_key, author).and_return(nil)
 
         expect {
-          SomeRelayable.from_xml(Nokogiri::XML::Document.parse(broken_xml).root)
-        }.to raise_error Entity::ValidationError, "invalid DiasporaFederation::SomeRelayable! missing 'parent_guid'."
+          Entities::SomeRelayable.new(hash).to_json
+        }.to raise_error Entities::Relayable::AuthorPrivateKeyNotFound
       end
-    end
 
-    context "fetch parent" do
-      before do
-        expect_callback(:fetch_public_key, author).and_return(author_pkey.public_key)
-        expect_callback(:fetch_public_key, remote_parent.author).and_return(parent_pkey.public_key)
+      it "doesn't set parent_author_signature if key isn't supplied" do
         expect_callback(:fetch_private_key, author).and_return(author_pkey)
-        expect_callback(:fetch_private_key, remote_parent.author).and_return(parent_pkey)
+        expect_callback(:fetch_private_key, local_parent.author).and_return(nil)
+
+        json = Entities::SomeRelayable.new(hash).to_json.to_json
+        expect(json).to include_json(entity_data: {parent_author_signature: ""})
       end
+    end
 
-      let(:xml) { SomeRelayable.new(hash).to_xml }
+    describe ".from_hash" do
+      let(:entity_class) { Entities::SomeRelayable }
 
-      it "fetches the parent from the backend" do
-        expect_callback(:fetch_related_entity, "Parent", parent_guid).and_return(remote_parent)
-        expect(Federation::Fetcher).not_to receive(:fetch_public)
+      context "parsing" do
+        before do
+          expect_callback(:fetch_related_entity, "Parent", parent_guid).and_return(remote_parent)
+          expect_callback(:fetch_public_key, author).and_return(author_pkey.public_key)
+          expect_callback(:fetch_public_key, remote_parent.author).and_return(parent_pkey.public_key)
+        end
+
+        context "when JSON properties are sorted and there is an unknown property" do
```

Comment:
Nit: Since this is `.from_hash` (and not JSON), maybe simply write `"when properties ..."`?